### PR TITLE
fix: Silently skip identical vocabulary entries during merge

### LINKED ISSUE #273: Merge produces 197 duplicate vocabulary warnings when both archives have standard vocabs
## Summary

When merging two archives that both embed standard vocabularies, `glx merge` reports ~230 "duplicate" warnings for every standard vocabulary entry. This buries the actually useful information (new entities and real entity ID conflicts).

### Root cause

The `mergeMap` helper in `go-glx/types.go` (line 544) only checks **key existence** — it does NOT compare values:

```go
if _, exists := dest[k]; exists {
    duplicates = append(duplicates, fmt.Sprintf("duplicate %s ID: %s", mapType, k))
}
```

Since both archives load the same standard vocabularies via `LoadStandardVocabulariesIntoGLX()`, every one of the 230 standard vocabulary entries triggers a "duplicate" message even though the values are identical.

### Vocabulary entry counts (230 total)

| Category | Count |
|----------|-------|
| event_types | 39 |
| participant_roles | 30 |
| person_properties | 24 |
| source_types | 19 |
| relationship_types | 17 |
| place_types | 16 |
| media_properties | 16 |
| repository_types | 10 |
| source_properties | 10 |
| repository_properties | 9 |
| citation_properties | 9 |
| media_types | 7 |
| place_properties | 7 |
| event_properties | 5 |
| confidence_levels | 4 |
| gender_types | 4 |
| relationship_properties | 4 |

### CLI UX best practices

Terraform plan is the gold standard: show only what changed, suppress identical items entirely, end with a summary count. The [CLI Guidelines (clig.dev)](https://clig.dev/) say: "The more irrelevant output you produce, the longer it's going to take the user to figure out what they did wrong."

### RDF set semantics as precedent

In RDF, a graph is a **set** of triples — duplicate triples collapse silently by definition. Standard vocabulary entries appearing in both archives are analogous: identical definitions for the same ID should merge silently, only conflicting definitions (same ID, different content) warrant a warning.

### Deep equality in Go

`reflect.DeepEqual` is the pragmatic choice for comparing vocabulary structs. The types are small, comparison happens once per merge, and there are no unexported fields. `google/go-cmp` is for tests only (may panic on unexported fields, performance is not a goal).

---

### Approach: Silent dedup for identical values, report only true conflicts

Modify `mergeMap` to compare values with `reflect.DeepEqual`. Identical entries are silently skipped. Only entries with the **same key but different values** are reported as conflicts. Return the count of identical skips so the CLI can show a summary.

### Changes

**1. `go-glx/types.go` — change `Merge()` return type and `mergeMap` logic**

Change `Merge()` signature from:
```go
func (g *GLXFile) Merge(other *GLXFile) []string
```
to:
```go
func (g *GLXFile) Merge(other *GLXFile) (conflicts []string, identicalSkipped int)
```

Change `mergeMap` to accept a `deepEqual bool` parameter (or create a variant). When `deepEqual` is true and `reflect.DeepEqual(dest[k], src[k])` is true, increment skip counter instead of appending to conflicts:

```go
func mergeMap[V any](mapType string, dest, src map[string]V, deepEqual bool) (conflicts []string, skipped int) {
    for k, v := range src {
        if existing, exists := dest[k]; exists {
            if deepEqual && reflect.DeepEqual(existing, v) {
                skipped++
            } else {
                conflicts = append(conflicts, fmt.Sprintf("duplicate %s ID: %s", mapType, k))
            }
        } else {
            dest[k] = v
        }
    }
    return conflicts, skipped
}
```

Use `deepEqual: false` for entity maps (entity ID collisions are always significant) and `deepEqual: true` for vocabulary/property maps (identical standard vocabs should be silent).

**2. `glx/merge_runner.go` — update output formatting**

Change the duplicates display from listing every entry to a tiered output:

```
  Conflicts (3 — skipped, destination kept):
    duplicate persons ID: person-1
    duplicate events ID: event-1
    duplicate sources ID: source-1

  230 identical vocabulary/property entries skipped
```

When there are no conflicts and only identical skips:
```
  No conflicts. 230 identical vocabulary entries skipped.
```

**3. `go-glx/types_merge_test.go` — update tests**

- Update existing tests to handle the new `(conflicts, skipped)` return values
- Add a test case for identical vocabulary entries (skipped == N, conflicts empty)
- Add a test case for conflicting vocabulary entries (same key, different label → appears in conflicts)

### What NOT to do

- **Don't add `--quiet-vocabs` flag** — silent dedup for identical items is the correct default, not a flag
- **Don't add `--verbose` flag** in this issue — that can be a follow-up for showing full details
- **Don't use `cmp.Equal` from go-cmp** — it's for tests only, may panic, performance is not a goal
- **Don't change entity duplicate behavior** — entity ID collisions are always worth reporting even if values happen to be identical

### Files to modify

1. `go-glx/types.go` — change `Merge()` and `mergeMap()` signatures and logic (~15 lines changed)
2. `go-glx/types_merge_test.go` — update test assertions for new return type (~20 lines changed)
3. `glx/merge_runner.go` — update `mergeArchivesInMemory` and output formatting (~15 lines changed)
4. `glx/merge_runner_test.go` — update test assertions (~5 lines changed)

### PR DESCRIPTION
**Pull request by isaacschepp**
## What and why

When merging two archives that both embed standard vocabularies, `glx merge` produced ~230 noisy "duplicate" warnings for every standard vocabulary entry, burying the useful information about new entities and real conflicts.

Now, identical vocabulary/property entries (same key + same value via `reflect.DeepEqual`) are silently skipped with a summary count. Only true conflicts (same key, different value) are reported.

Entity duplicates are still always reported regardless of value equality.

**Before:**
```
Duplicates (233 — skipped, destination kept):
  duplicate event_types ID: burial
  duplicate event_types ID: emigration
  ... (230 more lines)
  duplicate persons ID: person-1
```

**After:**
```
Conflicts (1 — skipped, destination kept):
  duplicate persons ID: person-1

230 identical vocabulary/property entries skipped
```

## Related issues

Closes #273

## Testing

- [x] 24 library-level merge tests pass (2 new: identical dedup, conflicting vocab)
- [x] CLI merge runner tests updated for new field names
- [x] Full test suite passes

## Breaking changes

`GLXFile.Merge()` signature changed from `[]string` to `([]string, int)`. This is a library API change but the project is pre-1.0 (beta).

### CODE REVIEW

**Review by noahbjohnson — APPROVED** [review]

File: go-glx/types.go [hunk 1]
```diff
@@ -548,6 +556,28 @@ func mergeMap[T any](mapType string, dest, src map[string]*T) []string {
...
+	if dest == nil || src == nil {
+		return nil, 0
+	}
+	for k, v := range src {
+		if existing, exists := dest[k]; exists {
+			if reflect.DeepEqual(existing, v) {
+				skipped++
+			} else {
+				conflicts = append(conflicts, fmt.Sprintf("conflict %s ID: %s", mapType, k))
+			}
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply go-glx/types.go:580]: Fixed: replaced `reflect.DeepEqual` with JSON-based comparison (`jsonEqual`) that treats nil maps and empty maps as equivalent.

[the remark this answers is not in the document]
Comment (isaacschepp) [reply go-glx/types.go]: Fixed: standardized all merge messages to use "conflict" prefix consistently — both entity collisions and vocabulary collisions now emit `conflict <type> ID: <id>`.

File: go-glx/types_merge_test.go [hunk 2]
```diff
@@ -600,8 +600,66 @@ func TestGLXFile_Merge_Metadata_ExistingEmptyDoesNotConflict(t *testing.T) {
...
+		PersonProperties: map[string]*PropertyDefinition{
+			"name": {ValueType: "integer"},
+		},
+	}
+
+	conflicts, skipped := g1.Merge(g2)
+	require.Len(t, conflicts, 2, "should report two conflicts")
+	require.Equal(t, 0, skipped, "no identical entries to skip")
+	require.Contains(t, conflicts[0], "conflict")
+	require.Contains(t, conflicts[1], "conflict")
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply go-glx/types_merge_test.go]: Fixed: assertions now check for specific entity type and ID (e.g. `event_types ID: birth`, `person_properties ID: name`) instead of just the word "conflict".

File: go-glx/types.go [hunk 3]
```diff
@@ -532,22 +542,55 @@ func (g *GLXFile) initMaps() {
...
+	}
+
+	return conflicts
+}
+
+// mergeMapDedup merges src into dest, silently skipping entries where the key
+// exists and the value is semantically equal. Only reports conflicts where the
+// same key maps to a different value. Uses JSON marshaling for comparison so
+// that nil maps and empty maps are treated as equivalent.
+func mergeMapDedup[T any](mapType string, dest, src map[string]*T) (conflicts []string, skipped int) {
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply go-glx/types.go:570]: Fixed: replaced `jsonEqual` with `yamlEqual` — YAML marshal with `omitempty` tags correctly treats nil and empty maps/slices as equivalent (both omitted), unlike JSON which emits `null` vs `{}`.

File: go-glx/types.go [hunk 4]
```diff
@@ -532,22 +542,55 @@ func (g *GLXFile) initMaps() {
...
+// jsonEqual compares two values by JSON-marshaling both, treating nil maps/slices
+// and empty maps/slices as equivalent.
+func jsonEqual(a, b any) bool {
+	aj, err1 := json.Marshal(a)
+	bj, err2 := json.Marshal(b)
+	if err1 != nil || err2 != nil {
+		return false
+	}
+
+	return string(aj) == string(bj)
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply go-glx/types.go]: Fixed: now uses `bytes.Equal(ay, by)` instead of string conversion.

File: glx/archive_io_test.go [hunk 5]
```diff
@@ -126,8 +126,8 @@ func TestLoadArchive(t *testing.T) {
 				if len(duplicates) != 1 {
 					return &testError{"expected 1 duplicate, got %d", []any{len(duplicates)}}
 				}
-				if duplicates[0] != "duplicate persons ID: person-1" {
-					return &testError{"expected duplicate to be 'duplicate persons ID: person-1', got %s", []any{duplicates[0]}}
+				if duplicates[0] != "conflict persons ID: person-1" {
+					return &testError{"expected duplicate to be 'conflict persons ID: person-1', got %s", []any{duplicates[0]}}
 				}
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply glx/archive_io_test.go:131]: Fixed: updated error messages from "duplicate" to "conflict".

File: go-glx/types.go [hunk 6]
```diff
@@ -391,60 +394,69 @@ type FieldDefinition struct {
...
 	g.validation = nil // invalidate cached validation since maps are being mutated
-	duplicates := make([]string, 0, 10)
+	conflicts = make([]string, 0, 10)
 
 	// Merge metadata (first one wins; report duplicate if both have content)
 	if other.ImportMetadata != nil && other.ImportMetadata.hasContent() {
 		if g.ImportMetadata != nil && g.ImportMetadata.hasContent() {
-			duplicates = append(duplicates, "duplicate metadata: metadata appears in multiple files")
+			conflicts = append(conflicts, "conflict metadata: metadata appears in multiple files")
 		} else {
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply go-glx/types.go:416]: Fixed: comment now says "report conflict if both have content".

File: go-glx/types.go [hunk 7]
```diff
@@ -391,60 +394,69 @@ type FieldDefinition struct {
...
-	duplicates = append(duplicates, mergeMap("event_properties", g.EventProperties, other.EventProperties)...)
-	duplicates = append(duplicates, mergeMap("relationship_properties", g.RelationshipProperties, other.RelationshipProperties)...)
-	duplicates = append(duplicates, mergeMap("place_properties", g.PlaceProperties, other.PlaceProperties)...)
-	duplicates = append(duplicates, mergeMap("media_properties", g.MediaProperties, other.MediaProperties)...)
-	duplicates = append(duplicates, mergeMap("repository_properties", g.RepositoryProperties, other.RepositoryProperties)...)
-	duplicates = append(duplicates, mergeMap("citation_properties", g.CitationProperties, other.CitationProperties)...)
-	duplicates = append(duplicates, mergeMap("source_properties", g.SourceProperties, other.SourceProperties)...)
-
-	return duplicates
+	// Merge entities — always report duplicates (entity ID collisions are significant)
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply go-glx/types.go]: Fixed: comment now says "always report conflicts".

File: go-glx/types_merge_test.go [hunk 8]
```diff
@@ -26,8 +27,8 @@ func TestGLXFile_Merge_EmptyFiles(t *testing.T) {
 	g1 := &GLXFile{}
 	g2 := &GLXFile{}
 
-	duplicates := g1.Merge(g2)
-	require.Empty(t, duplicates, "merging empty files should have no duplicates")
+	conflicts, _ := g1.Merge(g2)
+	require.Empty(t, conflicts, "merging empty files should have no duplicates")
 }
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply go-glx/types_merge_test.go:32]: Fixed: all test assertion messages updated from "duplicates" to "conflicts".

File: glx/merge_runner.go [hunk 9]
```diff
@@ -190,13 +192,17 @@ func mergeArchives(srcPath, destPath string, dryRun bool) error {
...
 
-	if len(result.Duplicates) > 0 {
-		fmt.Printf("\n  Duplicates (%d — skipped, destination kept):\n", len(result.Duplicates))
-		for _, d := range result.Duplicates {
+	if len(result.Conflicts) > 0 {
+		fmt.Printf("\n  Conflicts (%d — skipped, destination kept):\n", len(result.Conflicts))
+		for _, d := range result.Conflicts {
 			fmt.Printf("    %s\n", d)
 		}
 	}
```
[the remark this answers is not in the document]
Comment (isaacschepp) [reply glx/merge_runner.go:200]: merge_runner.go already uses the library output directly (which is now "conflict ..."). No change needed — the CLI output is consistent.
